go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,322 @@
+}
+
+type cachedConnDialer struct {
+	m            sync.Mutex
+	conns        map[string]*cachedConn
+	evict        []*cachedConn
+	evictSorted  bool
+	connWaitSema *sync2.Semaphore
+}
+
+var dialerStats = struct {
+	ConnReuse    *stats.Gauge
+	ConnNew      *stats.Gauge
+	DialTimeouts *stats.Gauge
+	DialTimings  *stats.Timings
+}{
+	ConnReuse:    stats.NewGauge("tabletmanagerclient_cachedconn_reuse", "number of times a call to dial() was able to reuse an existing connection"),
+	ConnNew:      stats.NewGauge("tabletmanagerclient_cachedconn_new", "number of times a call to dial() resulted in a dialing a new grpc clientconn"),
+	DialTimeouts: stats.NewGauge("tabletmanagerclient_cachedconn_dial_timeouts", "number of context timeouts during dial()"),
+	DialTimings:  stats.NewTimings("tabletmanagerclient_cachedconn_dialtimings", "timings for various dial paths", "path", "rlock_fast", "sema_fast", "sema_poll"),
+}
+
+// NewCachedConnClient returns a grpc Client that caches connections to the different tablets
+func NewCachedConnClient(capacity int) *Client {
+	dialer := &cachedConnDialer{
+		conns:        make(map[string]*cachedConn, capacity),
+		evict:        make([]*cachedConn, 0, capacity),
+		connWaitSema: sync2.NewSemaphore(capacity, 0),
+	}
+	return &Client{dialer}
+}
+
+var _ dialer = (*cachedConnDialer)(nil)
+
+func (dialer *cachedConnDialer) sortEvictionsLocked() {
+	if !dialer.evictSorted {
+		sort.Slice(dialer.evict, func(i, j int) bool {
+			left, right := dialer.evict[i], dialer.evict[j]
+			if left.refs == right.refs {
+				return right.lastAccessTime.After(left.lastAccessTime)
+			}
+			return right.refs > left.refs
+		})
+		dialer.evictSorted = true
+	}
+}
+
+func (dialer *cachedConnDialer) dial(ctx context.Context, tablet *topodatapb.Tablet) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	start := time.Now()
+	addr := getTabletAddr(tablet)
+
+	if client, closer, found, err := dialer.tryFromCache(addr, &dialer.m); found {
+		dialerStats.DialTimings.Add("rlock_fast", time.Since(start))
+		return client, closer, err
+	}
+
+	if dialer.connWaitSema.TryAcquire() {
+		defer func() {
+			dialerStats.DialTimings.Add("sema_fast", time.Since(start))
+		}()
+
+		// Check if another goroutine managed to dial a conn for the same addr
+		// while we were waiting for the write lock. This is identical to the
+		// read-lock section above.
+		if client, closer, found, err := dialer.tryFromCache(addr, &dialer.m); found {
+			dialer.connWaitSema.Release()
+			return client, closer, err
+		}
+		return dialer.newdial(ctx, addr)
+	}
+
+	defer func() {
+		dialerStats.DialTimings.Add("sema_poll", time.Since(start))
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			dialerStats.DialTimeouts.Add(1)
+			return nil, nil, ctx.Err()
+		default:
+			if client, closer, found, err := dialer.pollOnce(ctx, addr); found {
+				return client, closer, err
+			}
+		}
+	}
+}
+
+// tryFromCache tries to get a connection from the cache, performing a redial
```

Comment:
I'll clean up all these comments to reflect the actual state of the code tomorrow